### PR TITLE
[API] ModelClaim enforce minimum free HBM per GPU in placement

### LINKED ISSUE #2744: [Feature] Check free GPU memory before ModelClaim activation
### Problem
Previously, ModelClaim ranked warm Pods using their observed free HBM but did not check whether a Pod had enough memory to start the requested model. A Pod with a cached artifact could be selected even when its free HBM was too low.
If the engine encountered GPU OOM while loading the model, `Activate` could already have returned after spawning the process. The model remained non-routable while the runtime retried it on the same Pod. After the restart budget was exhausted, the runtime and ModelClaim reported a terminal failure; placement did not automatically try another Pod.
```mermaid
flowchart TD
    A["Controller selects a warm Pod"] --> B["Runtime Activate spawns the vLLM engine"]
    B --> C["Controller records Activating; route stays disabled"]
    C --> D["Engine loads model into GPU memory"]
    D --> E{"Enough GPU memory?"}
    E -- Yes --> F["Engine ready; ModelClaim becomes Active"]
    E -- No --> G["Engine exits; runtime cleans up and schedules restart"]
    G --> H{"Restart budget remaining?"}
    H -- Yes --> D
    H -- No --> I["Runtime reports failed"]
    I --> J["Controller marks ModelClaim Failed; route remains disabled"]
```

### Solution
Add optional spec.requiredHBMBytesPerGPU, expressed in bytes. When set, placement filters out Pods whose runtime snapshot has no GPU memory observation or whose available HBM is below the requested amount before applying artifact locality and load ranking. For multi-GPU placement, the existing snapshot summary uses the least free HBM across the required GPU group.
Claims that omit the field retain the previous placement behavior. If no Pod has confirmed sufficient capacity, the claim waits for a later reconciliation instead of calling Activate. The API schema, generated clients, tests, and ModelClaim documentation are updated.
This is a placement precheck, not a memory reservation. Concurrent activations and later KV usage can still change available HBM.

### Use Case

No response

### Proposed Solution

_No response_

### Area

Not sure

### PR DESCRIPTION
**Pull request by jiangxiaobin96**
fix [2744](https://github.com/vllm-project/aibrix/issues/2744)
Add an optional requiredHBMBytesPerGPU spec field that sets the minimum free GPU memory a candidate warm pod must report before the model is activated on it. Pods without a memory observation or with less free HBM than requested are skipped; when no candidate qualifies the claim stays Pending and reports NoMatchingPods with the unmet requirement.

- api: add the field, deepcopy, CRD schema, and apply configuration
- placement: filter candidates by confirmed free HBM when the limit is set
- docs: document the new spec field

## Pull Request Description
[Please provide a clear and concise description of your changes here]

## Related Issues
Resolves: #2744 

**Important: Before submitting, please complete the description above and review the checklist below.**

---

<details>
<summary><strong>Contribution Guidelines (Expand for Details)</strong></summary>

<p>We appreciate your contribution to aibrix! To ensure a smooth review process and maintain high code quality, please adhere to the following guidelines:</p>

<h3>Pull Request Title Format</h3>
<p>Your PR title should start with one of these prefixes to indicate the nature of the change:</p>
<ul>
    <li><code>[Bug]</code>: Corrections to existing functionality</li>
    <li><code>[CI]</code>: Changes to build process or CI pipeline</li>
    <li><code>[Docs]</code>: Updates or additions to documentation</li>
    <li><code>[API]</code>: Modifications to aibrix's API or interface</li>
    <li><code>[CLI]</code>: Changes or additions to the Command Line Interface</li>
    <li><code>[Misc]</code>: For changes not covered above (use sparingly)</li>
</ul>
<p><em>Note: For changes spanning multiple categories, use multiple prefixes in order of importance.</em></p>

<h3>Submission Checklist</h3>
<ul>
    <li>[ ] PR title includes appropriate prefix(es)</li>
    <li>[ ] Changes are clearly explained in the PR description</li>
    <li>[ ] New and existing tests pass successfully</li>
    <li>[ ] Code adheres to project style and best practices</li>
    <li>[ ] Documentation updated to reflect changes (if applicable)</li>
    <li>[ ] Thorough testing completed, no regressions introduced</li>
</ul>

<p>By submitting this PR, you confirm that you've read these guidelines and your changes align with the project's contribution standards.</p>

</details>

### CODE REVIEW

File: pkg/controller/modelclaim/placement.go [hunk 1]
```diff
@@ -126,6 +130,9 @@ func selectPodForActivationWithState(
 		}
 	}
 	if best == nil {
+		if requiredHBMBytesPerGPU > 0 {
+			return nil, fmt.Errorf("no candidate warm pod has confirmed free HBM of at least %d bytes per GPU", requiredHBMBytesPerGPU)
+		}
```
Comment (varungup90) [inline pkg/controller/modelclaim/placement.go:135]: If the candidate list is empty or every remaining pod is already hosting the model, this still returns the HBM-specific error whenever the spec field is set. That shows up on the `NoMatchingPods` condition and will send operators to debug GPU memory instead of selectors / replica packing.

Please only emit the HBM message when at least one not-already-on candidate was skipped for `!MemoryKnown` or `HBMFreeBytes < required`. Gemini’s suggestion of checking for an eligible candidate is the right idea; a `skippedForHBM` flag in the loop is even more precise.

Also worth a unit test for:
- no candidates + `requiredHBMBytesPerGPU > 0`
- all candidates already on + `requiredHBMBytesPerGPU > 0`

Both should keep the generic “no available candidate” error.

File: docs/source/features/modelclaim.rst [hunk 2]
```diff
@@ -235,6 +235,12 @@ The supported spec fields are:
      - Yes
      - Model artifact location, such as ``huggingface://``, ``s3://``, or
        ``gcs://``.
+   * - ``requiredHBMBytesPerGPU``
+     - No
+     - Minimum free GPU memory in bytes on each device before activation.
+       Include model weights, engine overhead, and serving headroom. When set,
+       Pods without a memory observation or enough free HBM are skipped.
+       When omitted, placement uses its existing best-effort ranking.
```
Comment (varungup90) [inline docs/source/features/modelclaim.rst:243]: Please make this match the snapshot summary and the rest of the page:

- For TP=1, `HBMFreeBytes` is the **largest** single-device slot, not min across every GPU. “On each device” is only accurate for TP/PP > 1 (least-free rank in the group). See `placementStateFromSnapshot`.
- This is a placement **precheck**, not a reservation. Concurrent Activate calls can still race.
- The Observability section later still says HBM “is not a hard admission or reservation signal.” That is now false when this field is set; please update it in this PR.
- Troubleshooting currently only covers “zero candidates.” Add the case where `status.candidates > 0` but the claim stays Pending with `NoMatchingPods` because no pod has confirmed free HBM.

File: api/model/v1alpha1/modelclaim_types.go [hunk 3]
```diff
@@ -45,6 +45,13 @@ type ModelClaimSpec struct {
 	// +kubebuilder:validation:Required
 	ArtifactURL string `json:"artifactURL,omitempty"`
 
+	// RequiredHBMBytesPerGPU is the minimum free GPU memory required on each
+	// device before activating this model. Include weights, engine overhead,
+	// and serving headroom. When omitted, placement does not enforce capacity.
+	// +optional
+	// +kubebuilder:validation:Minimum=1
+	RequiredHBMBytesPerGPU *int64 `json:"requiredHBMBytesPerGPU,omitempty"`
```
Comment (varungup90) [inline api/model/v1alpha1/modelclaim_types.go:53]: “On each device” overstates the TP=1 behavior. Placement reuses `PodPlacementState.HBMFreeBytes`, which is the largest available single-device slot when parallelism is 1 and the least-free rank when TP/PP > 1.

Please spell that out here (and that unset means no capacity filter). Operators will treat this godoc as the API contract.

File: pkg/controller/modelclaim/placement.go [hunk 4]
```diff
@@ -104,6 +104,7 @@ func selectPodForActivationWithState(
 	model string,
 	locality LocalityProvider,
 	states map[string]PodPlacementState,
+	requiredHBMBytesPerGPU int64,
```
Comment (varungup90) [inline pkg/controller/modelclaim/placement.go:107]: The comment above this function still says missing runtime state “simply falls back to the existing deterministic placement behavior.” With `requiredHBMBytesPerGPU > 0`, missing `MemoryKnown` **excludes** the pod. Please update the comment so the hard filter is documented next to the new parameter.

**Review by varungup90 — COMMENTED** [review]
## Review

The HBM filter matches #2744 and the existing snapshot summary (max slot for TP=1, min-free rank for TP>1). `TestReconcilePlacementRequiresHBM` covers the original bug: a cached pod under the limit loses to a pod with confirmed capacity.

Please address the two operator-facing issues below before merge:

1. **Docs are now internally inconsistent.** The new spec field is a hard admission filter, but the Observability section still says HBM “is not a hard admission or reservation signal.” Troubleshooting also only covers zero candidates, not `Candidates > 0` + `NoMatchingPods` because of HBM.
2. **HBM error attribution is too broad.** If there are no eligible candidates at all, the claim still reports a missing-HBM error. Gemini called this out on `placement.go`; please only use that message when a pod was actually skipped for HBM, and add tests for empty / already-on candidates.

Nits (non-blocking): fill `Resolves: #2744` so the `triage/needs-information` label can drop; the repo title convention wants `[API][Docs]` rather than `feat(modelclaim):`. `resource.Quantity` (`40Gi`) would be easier for users than raw bytes, but that can be a follow-up while this is still v1alpha1.

**Comment by jiangxiaobin96** [comment]
@varungup90 Please help to review again.

File: config/crd/model/model.aibrix.ai_modelclaims.yaml [hunk 5]
```diff
@@ -96,6 +96,15 @@ spec:
                 maximum: 1
                 minimum: 1
                 type: integer
+              requiredHBMBytesPerGPU:
```
Comment (bolubo) [inline config/crd/model/model.aibrix.ai_modelclaims.yaml:99]: The `Verify Generated Code` job fails on `61c6ba00` (the same failure is on `2b55e89f`): `hack/verify-crd-sync.sh` reports that `dist/chart/crds/model.aibrix.ai_modelclaims.yaml` differs from the generated `config/crd/bases/model.aibrix.ai_modelclaims.yaml` ([job log](https://github.com/vllm-project/aibrix/actions/runs/35516317260/job/106092754335)). The Helm chart copy is missing the entire `requiredHBMBytesPerGPU` schema block, so Helm installs would silently prune the field. That chart copy landed on main in #2701 after this branch point; please rebase onto main, run `make sync-crds-to-helm`, and commit `dist/chart/crds/model.aibrix.ai_modelclaims.yaml`.
